docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,232 @@
+## Non Goals
+
+Non goals represent capabilities that are not going to be done in the initial release of modules but may come in later versions.  
+
+* Add additional capabilities to load strings
+* Any type of versioning
+* Any user interface work beyond ensuring it works as the UI currently does
+* Any sort of security for modules and what modules are allowed
+* Any sort of metadata
+
+
+
+### Add additional capabilities to load strings
+
+Modules will not care about the source of the string that represents the river syntax, nor will modules have any inherent reload semantics. The component supplying the string will be responsible for the source and will notify the module when the input changes so that it can utilize the new river configuration.
+
+### Any type of versioning
+
+Modules will not contain any sort of versioning nor will check for compatibility outside the normal river checks.
+
+### Any user interface work beyond ensuring it works as the UI currently does
+
+Users will not be able to drill into modules, they will be represented as any other normal component. 
+
+## Example
+
+```river
+// module
+argument "password" {
+    optional = false
+    comment = "password for mysql"
+}
+
+argument "username" {
+    optional = false
+    comment = "username for mysql"
+}
+
+export "targets" {
+    comment = "targets for the integration"
+    value = integrations.mysql.server1.targets
+}
+
+integrations.mysql "server1" {
+    username = argument.username.value
+    password = argument.password.value
+}
+
+```
+
+```river
+// parent
+
+local.file "mysql" {
+    filename = "/test/mysql.river"
+}
+
+module.string "mysql" {
+    content = local.file.mysql.content
+    arguments = {
+        {
+            "password" = PASSWORD,
+            "username" = USERNAME,
+        }
+    }
+}
+
+prometheus.scrape "scraper" {
+    targets = module.string.mysql.exports.targets
+}
+
+```
+
+## Limitations
+
+* A module cannot directly or indirectly load itself, this will not be enforced by the system
+* Singleton components are not supported at this time. Example [node_exporter](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.integration.node_exporter/).
+* Modules will not prevent competing resources, such as starting a server on the same port
+* [Configuration blocks](https://grafana.com/docs/agent/latest/flow/reference/config-blocks/#configuration-blocks) will not be supported. 
+* Names of arguments and exports within a module must be unique across that module.
+
+## Proposal
+
+Add the ability to load `modules` as subgraphs to the primary `graph`. Modules may call other modules within a reasonable stack size depth. Modules are represented as a river string that is interpreted with a defined set of arguments and exports.
+
+The initial component will be `module.string` that will load a single module. Internally these modules will be namespaced so they cannot affect children or parent graphs except via arguments and exports.
```

Comment:
If a module reports certain metrics, then what would happen if we import the module more than once? I guess that there will be a label to indicate the fully qualified module name which contains these metrics?

Reply:
The namespace will include the parent labels, so that will prevent duplication. 

```river
module.string "modulo" {
    content = "some string that's valid"
}

module.string "samestringmodule" {
  content  = "some strings that's valid"
}
```

In the above, the components for `module.string.modulo` will be prefixed with `module.string.modulo` which trickles down to `data-agent` and metric nanes. Components under `module.string.samestringmodule` would be prefixed with `module.string.samestringmodule`.